[feat]support colored material icons

xtliu97 opened the pull request:
## Describe your changes
support colored material icons

<img width="389" alt="image" src="https://github.com/user-attachments/assets/0dd10130-7feb-4cea-90a0-00018ccebcd5" />


## GitHub Issue Link (if applicable)
https://github.com/streamlit/streamlit/issues/11426

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

xtliu97 committed:
[feat]support colored material icons

jrieke commented:
Just tried this out and it works great! The only thing that doesn't seem to work is `rainbow` in `st.warning`, `st.info`, `st.error`, and `st.success`:

![Screenshot 2025-05-23 at 22 03 39](https://github.com/user-attachments/assets/ed68e898-810b-404f-84ac-8aa15b449dc6)

Fine with merging this from product side as soon as that is fixed and it passes code review!

xtliu97 committed:
[fix] add  prop support to MaterialFontIcon

xtliu97 commented:
@jrieke  Thanks for the feedback! I've just pushed a fix for the rainbow color in status elements and it now works.
<img width="771" alt="image" src="https://github.com/user-attachments/assets/4f4253e1-7d72-4fb8-96d2-6cff1d40df75" />

xtliu97 committed:
fix for lint

sfc-gh-lwilby commented:
@xtliu97 thank you for submitting this PR. You have some CI failures that need to be fixed, are you able to take a look?

xtliu97 commented:
> @xtliu97 thank you for submitting this PR. You have some CI failures that need to be fixed, are you able to take a look?

Thanks for the reminder! I've fixed the issues causing the CI failures. Could you please help re-run the workflow?

sfc-gh-lwilby commented on frontend/lib/src/components/shared/Icon/DynamicIcon.tsx:
```diff
@@ -97,15 +113,53 @@ const DynamicIconDispatcher = ({
+  const theme: EmotionTheme = useTheme()
+  const themeMode = hasLightBackgroundColor(theme) ? "light" : "dark"
+  const colorMapping = createColorMapping(theme)
+  const style = parsedColor ? colorMapping.get(parsedColor) : {}
```
Not super familiar with this code so far, but this seems wrong. 

But @xtliu97 , what about `props.color`? It looks like here it will be overridden by the color passed in the icon. Is that the default theme color?

sfc-gh-lwilby commented on frontend/lib/src/components/shared/Icon/Material/MaterialFontIcon.tsx:
```diff
@@ -31,6 +31,7 @@ interface MaterialIconProps {
   margin?: string
   padding?: string
   testid?: string
+  style?: React.CSSProperties
```
This name seems too broad a name, we should pick something more descriptive. I expected it to be called `color`, but I see that is already included. How do `color` above and the new color prop here interact?

xtliu97 replied:
The style prop here refers to the standard inline style object used in React components, and I think I should remove it here.

sfc-gh-lwilby replied:
We tend to prefer using more specific names instead of passing the styles via a general style prop. Similar to how `padding` and `margin` are handled. That's what I meant by this is too broad.

xtliu97 replied:
Got it. I have changed it to specific prop names.

sfc-gh-lwilby commented on lib/tests/streamlit/elements/button_test.py:
```diff
@@ -57,6 +59,22 @@ def test_material_icon(self):
+        c = self.get_delta_from_queue().new_element.button
+        assert c.icon == ":red[:material/thumb_up:]"
+
+    def test_invalid_colored_material_icon(self):
```
What if the color isn't in the expected position in the icon string?

xtliu97 replied:
It will raise a `StreamlitAPIException` tells that it is not a valid emoji. I will add an another test case here.

sfc-gh-lwilby commented on frontend/lib/src/components/shared/BaseButton/DynamicButtonLabel.tsx:
```diff
@@ -33,7 +33,9 @@ export const DynamicButtonLabel = ({
-  const isMaterialIcon = icon?.startsWith(":material")
+  const isMaterialIcon =
+    icon?.startsWith(":material") ||
+    icon?.match(/^:[^[]+\[:material\/[^\]]+\]$/)
```
Could you use * instead of + and cover both conditionals with the regex?

sfc-gh-lwilby commented:
I think this PR needs more test coverage to give e2e coverage, currently we are only testing the backend. 

Can you add some playwright e2e tests?

It would be nice to add some other frontend test coverage, particularly with all of the regex parsing involved.

xtliu97 committed:
[fix] merge regex of judging colored material icon and add test for invalid emoji

xtliu97 committed:
[fix] use specific prop names when passing icon colors

xtliu97 committed:
add e2e-playwright test for colored material icon in button

xtliu97 committed:
fix colormap type

sfc-gh-lwilby commented on frontend/lib/src/components/shared/BaseButton/DynamicButtonLabel.tsx:
```diff
@@ -33,7 +33,9 @@ export const DynamicButtonLabel = ({
   useSmallerFont = false,
 }: DynamicButtonLabelProps): React.ReactElement | null => {
-  const isMaterialIcon = icon?.startsWith(":material")
+  const isMaterialIcon = icon?.match(
```
@xtliu97 I agree with this, even though it is very short if this was a util function, then you could also add some tests. That will help both with readability/understanding and also just validate the regex is what we want.

sfc-gh-lwilby commented:
@xtliu97 before this PR can be merged, please take a look at the agent review comments. I glanced through them and it looks like they have some good feedback. 

Also, it looks like some of the snapshots you added need to be updated.

jrieke commented:
@xtliu97 You're still interested in finishing this PR? I'd love to see this feature become reality! :)

sfc-gh-lwilby committed:
Revert icon sizing changes to keep PR focused on colored icons feature

- Remove dynamic icon sizing logic that affected all material icons
- Keep default 'lg' size for all icons (same as develop branch)
- This change was out of scope for the colored material icons feature

sfc-gh-lwilby committed:
Add clarifying comment about color prop behavior in DynamicIcon

- The color prop serves as a fallback when no color is specified in the icon string
- Parsed colors from icon strings (:red[:material/home:]) override props.color
- This addresses PR feedback about props.color interaction

sfc-gh-lwilby commented on frontend/lib/src/components/shared/Icon/DynamicIcon.tsx:
```diff
@@ -43,6 +52,20 @@ export function parseIconPackEntry(iconName: string): IconPackEntry {
+/**
+ * Parses the icon value to extract color and icon name.
+ */
+function parseColorAndIcon(iconValue: string): {
```
These functions might better belong in a util file.

sfc-gh-lwilby committed:
Improve docstrings for colored icon validation functions

- Add comprehensive Numpydoc-style docstring to validate_icon_or_emoji
- Add comprehensive Numpydoc-style docstring to validate_color
- Document parameters, return values, and exceptions
- Addresses PR feedback about documentation quality

sfc-gh-lwilby committed:
Add test coverage for colored material icons in chat messages

- Add unit test for regular material icon avatar
- Add unit test for colored material icon avatar
- Add E2E test examples with :red and :rainbow colored icons
- Update test counts and descriptive names for snapshots
- Addresses PR feedback about test coverage completeness

sfc-gh-lwilby committed:
Simplify colored icon implementation to use consistent style prop

- Use React.CSSProperties for all color styles (regular and rainbow)
- Apply all colors through single 'style' prop instead of mixed prop/style
- Remove extraColorStyle logic from MaterialFontIcon
- Merge user-provided styles with color styles properly
- No visual changes, just cleaner implementation

This addresses PR feedback about code complexity and type consistency.

jrieke commented:
Closing since this seems to be stale. Feel free to reopen if you continue working on this!
